### PR TITLE
ci: Use honeycomb to trace builds 

### PR DESCRIPTION
**Pull request by jhchabran**
This PR integrates honeycomb's [buildevents](https://github.com/honeycombio/buildevents) in our pipeline, enabling us to trace build and to get an overview of the timings. It goes down to the command level. 

Associated infra PR: https://github.com/sourcegraph/infrastructure/pull/2997

<details>
  <summary>Dashboard preview</summary>
image
</details> 

<details>
  <summary>Trace preview</summary>
image
</details> 

<details>
  <summary>Annoatation preview</summary>
image
</details> 

TODO: 

- [x] Draft a dashboard on Honeycomb
- [x] Send events from a PR 
- [x] Send events from a PR with the full main dry run (including E2E agents).
- [x] Extract the `buildevents` binary from the `post-checkout` hook to avoid downloading it everytime. 
- [x] Merge the infra PR and deploy the new agents

### CODE REVIEW

File: .buildkite/hooks/post-command [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env bash
+
+(
```
Comment (jhchabran) [inline .buildkite/hooks/post-command:3]: The `()` prevents the variable export to escape the scope of that command while still allowing underlying shells to see it. We wouldn't want to start tracing things by accident don't we 😄?

File: enterprise/dev/ci/internal/buildkite/buildkite.go [hunk 2]
```diff
@@ -198,7 +199,13 @@ type StepOpt func(step *Step)
 
 func Cmd(command string) StepOpt {
 	return func(step *Step) {
-		step.Command = append(step.Command, command)
+		// step.Command = append(step.Command, command)
```
Comment (jhchabran) [inline enterprise/dev/ci/internal/buildkite/buildkite.go:202]: TODO drop

File: .buildkite/hooks/post-checkout [hunk 3]
```diff
@@ -9,3 +9,12 @@ echo "Running git fetch... done"
 echo "install asdf plugins"
 asdf install
 echo "done installing"
+
+# TODO Temporary download so I can test this branch without its companion on sourcegraph/infrastructure
+echo "Fetching buildevents"
+curl -L -o buildevents https://github.com/honeycombio/buildevents/releases/latest/download/buildevents-linux-amd64
+chmod 755 buildevents
+# mv buildevents /usr/local/bin
```
Comment (jhchabran) [inline .buildkite/hooks/post-checkout:17]: TODO: Don't forget to update this after deploying the new agents.

File: enterprise/dev/ci/internal/ci/operations.go [hunk 4]
```diff
@@ -688,3 +688,11 @@ func publishExecutorDockerMirror(version string) operations.Operation {
 		pipeline.AddStep(":packer: :white_check_mark: docker registry mirror image", stepOpts...)
 	}
 }
+
+func buildeventUploadTrace() operations.Operation {
```
Comment (bobheadxi) [inline enterprise/dev/ci/internal/ci/operations.go:692]: nit - more inline with the script and verb first reads a little smoother:

```suggestion
func uploadBuildeventTrace() operations.Operation {
```

File: enterprise/dev/ci/internal/buildkite/buildkite.go [hunk 5]
```diff
@@ -198,7 +199,13 @@ type StepOpt func(step *Step)
...
+
+		// BUILDKITE_STEP_ID is prefixed by $$ so it's not interpolated at pipeline generation time
+		// but instead when the step is executed.
+		// TODO
+		tracedCmd := fmt.Sprintf("./tr %s", command)
+		step.Command = append(step.Command, tracedCmd)
```
Comment (bobheadxi) [inline enterprise/dev/ci/internal/buildkite/buildkite.go:208]: Love this, exactly the sort of powerful stuff having a pipeline generator enables us to do!

**Comment by jhchabran** [comment]
Successful run https://buildkite.com/sourcegraph/sourcegraph/builds/124860

File: enterprise/dev/ci/internal/buildkite/buildkite.go (hunk 5 above)
Comment (jhchabran) [reply enterprise/dev/ci/internal/buildkite/buildkite.go:208]: Yes, this is awesome 😊

**Comment by jhchabran** [comment]
FYI, the whole thing works, I'm waiting for them to merge my PR so we don't have to download my forked version from my dropbox (🐗).

**Comment by jhchabran** [comment]
https://github.com/honeycombio/buildevents/releases/tag/v0.8.0 is out, fixing the code

**Comment by jhchabran** [comment]
Will do a final test tomorrow, so we can get this merged and collect some data to compare with the stateless agents as we refine them 😊 

@bobheadxi do you think you could do a final review in between so I can merge this if and only if the final test is successful ?

File: enterprise/dev/upload-buildevent-report.sh [hunk 6]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
```
Comment (bobheadxi) [inline enterprise/dev/upload-buildevent-report.sh:1]: Maybe this script should be in https://github.com/sourcegraph/sourcegraph/blob/c36f1b0bf2fcd94bcf9c9d1f4e8218f5a834f380/enterprise/dev/ci/scripts ?

File: enterprise/dev/trace-command.sh [hunk 7]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
```
Comment (bobheadxi) [inline enterprise/dev/trace-command.sh:1]: Same with this script, maybe move to https://github.com/sourcegraph/sourcegraph/blob/c36f1b0bf2fcd94bcf9c9d1f4e8218f5a834f380/enterprise/dev/ci/scripts

File: enterprise/dev/ci/internal/buildkite/buildkite.go [hunk 8]
```diff
@@ -198,7 +199,10 @@ type StepOpt func(step *Step)
 
 func Cmd(command string) StepOpt {
 	return func(step *Step) {
-		step.Command = append(step.Command, command)
+		// BUILDKITE_STEP_ID is prefixed by $$ so it's not interpolated at pipeline generation time
+		// but instead when the step is executed.
+		tracedCmd := fmt.Sprintf("./tr %s", command)
```
Comment (bobheadxi) [inline enterprise/dev/ci/internal/buildkite/buildkite.go:204]: I'm not sure I follow what this comment is referring to, could you clarify?

**Review by bobheadxi — APPROVED** [review]
LGTM, left some minor comments! It is a lot of bash magic though, but I suppose that is inevitable 😬 Excited to have this!

File: enterprise/dev/ci/internal/buildkite/buildkite.go (hunk 8 above)
Comment (jhchabran) [reply enterprise/dev/ci/internal/buildkite/buildkite.go:204]: It's a leftover. Thanks!

File: .buildkite/hooks/pre-command [hunk 9]
```diff
@@ -25,3 +25,17 @@ for file in "${TOOL_VERSION_FILES[@]}"; do
...
+
+# HoneyComb's buildevent plumbing.
+# -------------------------------
+
+# Record start time if we need to exit
+BUILD_START_TIME=$(curl -H "Authorization: Bearer $BUILDKITE_API_TOKEN" "https://api.buildkite.com/v2/organizations/$BUILDKITE_ORGANIZATION_SLUG/pipelines/$BUILDKITE_PIPELINE_SLUG/builds/$BUILDKITE_BUILD_NUMBER/" | jq -r .started_at)
+export BUILD_START_TIME
+
+# Convert to UTC & Epoch
+BUILD_START_TIME=$(TZ=UTC date -d "$BUILD_START_TIME" +'%s')
```
[the remark this answers is not in the document]
Comment (jhchabran) [reply .buildkite/hooks/pre-command:37]: Oh good catch. You beat `shellcheck` there :D

**Comment by jhchabran** [comment]
Final QA: https://buildkite.com/sourcegraph/sourcegraph/builds/125719 🎉

**Pull request merged**